dask/dataframe/utils.py

```diff
@@ -343,7 +343,13 @@ def _nonempty_index(idx):
         return pd.CategoricalIndex(data, categories=cats,
                                    ordered=idx.ordered, name=idx.name)
     elif typ is pd.MultiIndex:
-        levels = [_nonempty_index(i) for i in idx.levels]
```

Comment:
It seems like there are two conflicting constraints in the nonempty_index for CategoricalIndex

1. A desire to use the real categories for the `.categories`
2. A desire for every non_empty index to be length 2.

If given a length-0 or length-1 CategoricalIndex, we can't satisfy both at once. I'd *like* to solve this issue there, instead of in the MultiIndex part, but that may not be possible. If not, then this approach seems fine.

Reply:
I tried to fix it for the CategoricalIndex first but ran into the issues you mentioned already. I was also a bit of confused why the MultiIndex requires unique data in the first place and wasn't sure if this is a pandas issue. From your comment, I guess this is intended behavior?

Reply:
I forgot to link to the pandas issue, https://github.com/pandas-dev/pandas/issues/17464. This was a deliberate change by pandas. Levels of a MI should be unique.

How about this... 

```python
         if len(idx.categories) == 0:
             data = _nonempty_index(idx.categories)
             cats = None
         elif len(idx.categories) == 1:
             data = _nonempty_index([idx.categories[0], None])
             cats = idx.categories
         else:
             data = idx.categories[:2]
             cats = idx.categories
```

Does that work?

Reply:
1. I have to pass the list as an index, e.g. `_nonempty_index(pd.Index([idx.categories[0], None]))` for the second case to work out. However, this would ignore the type of the input (not sure if it matters, though. The `data` types are not tested). Another alternative which preserves the type of `data` would be

```python
        if len(idx.categories) == 0:
            data = _nonempty_index(idx.categories)
            cats = None
        else:
            data = _nonempty_index(_nonempty_index(idx.categories))
            cats = idx.categories
```

2. Both solutions ignore the special case of `__UNKNOWN_CATEGORIES__` which ends up as `CategoricalIndex([nan, nan], categories=['__UNKNOWN_CATEGORIES__'], ordered=True, name='foo', dtype='category')`. I assume it is pandas setting the data purposely to nan for unknown categories and this is intended behavior, isn't it?